llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -3109,6 +3109,33 @@ Value *LibCallSimplifier::optimizeRemquo(CallInst *CI, IRBuilderBase &B) {
   return ConstantFP::get(CI->getType(), Rem);
 }
 
+/// Constant folds fdim
+Value *LibCallSimplifier::optimizeFdim(CallInst *CI, IRBuilderBase &B) {
+  const APFloat *X, *Y;
+  // Check if both values are constants
+  if (!match(CI->getArgOperand(0), m_APFloat(X)) ||
+      !match(CI->getArgOperand(1), m_APFloat(Y)))
+    return nullptr;
+  // If either argument is NaN, NaN is returned
+  if (X->isNaN() || Y->isNaN())
+    return ConstantFP::getQNaN(CI->getType());
+
+  IRBuilderBase::FastMathFlagGuard Guard(B);
+  FastMathFlags FMF = CI->getFastMathFlags();
+  // set no-NaN fast-math-flag as we already checked for NaN for both operands
+  FMF.setNoNaNs();
+  // set no-signed-zeroes as fdim will never return -0.0
+  FMF.setNoSignedZeros();
+  B.setFastMathFlags(FMF);
+  // fdim is equivalent to fmax(x - y, 0), except for the NaN handling
```

Comment:
I think you can only do this if you know that math errno isn't being used. fdim sets errno on range errors, but the fmax intrinsic does not. An example that shows this is fdim(1e308, -1e308);

Reply:
correct, I am thinking of abandoning fdim -> fmax, as this PR is for constant propogation
and i thougth fdim->fmax would mean i won't have to handle corner cases manually

Reply:
I think it is better to just transform into the more understood ops

Reply:
I agree it would be better to transform to fsub+fmax, if you can be sure errno doesn't need to be set. I suppose the "memory(none)" attribute establishes that, as you noted in another commented. A case could be made that the 'afn' fast-math flag would also allow it.

Reply:
ok, so i added checks so the fold is done only when `fsub` return ok and fdim has `memory(none)` attribute

poison and undefs are not getting propogated, will have to look into that